### PR TITLE
Convert Cuboid2D to/from KITTI 3D data

### PR DESCRIPTION
**Pull request by itrushkin**
### Summary
CVS-151427

#### New features
- New `Cuboid2D` methods:
  - `Cuboid2D.from_3d(dimensions, location, rotation_y, P, Tr_velo_to_cam)`: Creates a Cuboid2D object from KITTI 3D bbox annotation data. Matrix `P` (`P2` in Kitti format context) is a 3x4 projection matrix in the left color camera coordinate system. Matrix `Tr_velo_to_cam` is a 3x4 projection matrix between LiDAR and camera coordinate systems.
  - `cuboid_2d.to_3d(P_inv)`: Reconstructs approximate KITTI 3D bbox annotation data (`dimensions`, `location` and `rotation_y`) from 2D projection coordinates. `P_inv` matrix is a [pseudo-inverse](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse) of camera-to-image projection matrix.

### How to test
See unit test changes

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [x] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: src/datumaro/components/annotation.py [hunk 1]
```diff
@@ -1393,11 +1394,166 @@ class Cuboid2D(Annotation):
...
+        b2 = y3 - y1
+        c2 = z3 - z1
+        a = b1 * c2 - b2 * c1
+        b = a2 * c1 - a1 * c2
+        c = a1 * b2 - b1 * a2
+        d = -a * x1 - b * y1 - c * z1
+        return np.array([a, b, c, d])
+
+    @staticmethod
+    def _get_denorm(Tr_velo_to_cam_homo):
```
Comment (sooahleex) [inline src/datumaro/components/annotation.py:1427]: Just a question, what is the meaning of `Tr_velo_to_cam_homo`?

File: src/datumaro/components/annotation.py [hunk 2]
```diff
@@ -1372,19 +1373,19 @@ class Cuboid2D(Annotation):
...
-      6---7
+      2---3
      /|  /|
-    5-+-8 |
-    | 2 + 3
+    1-+-4 |
+    | 5 + 6
     |/  |/
-    1---4
+    8---7
```
Comment (sooahleex) [inline src/datumaro/components/annotation.py:1381]: I understood the overall structure, is there any reason to change the bottom and top face?

**Review by sooahleex — COMMENTED** [review]
Could you update the documents and explanation of this feature in this pr too?

File: src/datumaro/components/annotation.py (hunk 1 above)
Comment (itrushkin) [reply src/datumaro/components/annotation.py:1427]: Calibration matrix `Tr_velo_to_cam` has a shape of 3 x 4. To project 3D points to the 2D plane, we need to have [homogeneous coordinates](https://en.wikipedia.org/wiki/Homogeneous_coordinates) where each point is represented as a vector with 1 additional dimension.

File: src/datumaro/components/annotation.py (hunk 2 above)
Comment (itrushkin) [reply src/datumaro/components/annotation.py:1381]: I aligned the order of points with Kitti format which describes top face first.

**Comment by itrushkin** [comment]
> Could you update the documents and explanation of this feature in this pr too?

@sooahleex, documentation is added in [657bf1b](https://github.com/openvinotoolkit/datumaro/pull/1639/commits/657bf1ba87275391ea0bdbab24806e4c29da7b15). PR description is extended as well.

File: src/datumaro/components/annotation.py (hunk 1 above)
Comment (wonjuleee) [reply src/datumaro/components/annotation.py:1427]: This is the projection matrix between Velodyne LiDAR to Camera, where LiDAR contains 4 dimensions ([X, Y, Z, 1]) and Camera contains 3 dimensions ([u, v, 1]). `velo` stands for Velodyne :)

**Review by wonjuleee — APPROVED** [review]

**Review by sooahleex — APPROVED** [review]

**Pull request merged**